src/main/java/com/iota/iri/service/snapshot/impl/SnapshotServiceImpl.java

```diff
@@ -653,63 +527,134 @@ private void processOldSolidEntryPoints(Tangle tangle, SnapshotProvider snapshot
-                );
-
-                solidEntryPoints.put(currentMilestone.getHash(), targetMilestone.index());
-
-                nextMilestone = MilestoneViewModel.findClosestPrevMilestone(tangle, currentMilestone.index(),
-                        snapshotProvider.getInitialSnapshot().getIndex());
-
-                progressLogger.progress();
+            Map<Hash, Integer> entryPoints = getSolidEntryPoints(targetMilestone.index(), progressLogger);
+            
+            // Only is null when we are shutting down / Thread interrupted
+            if (entryPoints == null) {
+                throw new SnapshotException("No solid entrypoint generation due to thread interruption");
             }
-
+            
+            solidEntryPoints.putAll(entryPoints);
             progressLogger.finish();
         } catch (Exception e) {
             progressLogger.abort(e);
-
             throw new SnapshotException("could not generate the solid entry points for " + targetMilestone, e);
         }
     }
+    
+    /**
+     * Generates entrypoints based on target index down to the maximum depth of the node
+     * 
+     * @param targetIndex The milestone index we target to generate entrypoints until.
+     * @param progressLogger The logger we use to write progress of entrypoint generation
+     * @return a map of entrypoints or <code>null</code> when we were interrupted
+     * @throws Exception When we fail to get entry points due to errors generally caused by db interaction
+     */
+    private Map<Hash, Integer> getSolidEntryPoints(int targetIndex, ProgressLogger progressLogger) throws Exception {
+        Map<Hash, Integer> solidEntryPoints = new HashMap<>();
+        solidEntryPoints.put(Hash.NULL_HASH, targetIndex);
+        log.info("Generating entrypoints for {}", targetIndex);
+        
+        // Co back a but below the milestone. Limited to maxDepth or genisis
+        int startIndex = Math.max(snapshotProvider.getInitialSnapshot().getIndex(), targetIndex - config.getMaxDepth()
+                ) + 1; // cant start at last snapshot now can we, could be 0!
+        
+        progressLogger.start(startIndex);
+        
+        // Iterate from a reasonable old milestone to the target index to check for solid entry points
+        for (int milestoneIndex = startIndex; milestoneIndex <= targetIndex; milestoneIndex++) {
+            if (Thread.interrupted()) {
+                return null;
+            }
+            
+            MilestoneViewModel milestone = MilestoneViewModel.get(tangle, milestoneIndex);
+            if (milestone == null) {
+                log.warn("Failed to find milestone {} during entry point analyzation", milestoneIndex);
+                return null;
+            }
+            
+            List<Hash> approvees = getMilestoneApprovees(milestoneIndex, milestone);
+            for (Hash approvee : approvees) {
+                if (Thread.interrupted()) {
+                    return null;
```

Comment:
usually when one resets the interrupted status (you are implicitly doing it here) , one throws an `InterruptedException` (and resets the interrupted status when you catch the error)... or at least one doesn't reset the interruption status..

The reason is that technically when you look at this method you are only aware of its scope and you are not aware of whether this thread will die later on. Imagine a case where the thread was interrupted but continues to work, and on the next check

You can argue that in this specific context it will work.. it may break if the code changes but this will probably not happen. Even though this may be correct, I ask you: why do things the wrong way to begin with?

Reply:
see comment above